web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx

```diff
@@ -124,9 +132,60 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
           visible={props.visible}
           unsanitizedFontFamily={unsanitizedTerminalFontFamily}
           fontSize={terminalFontSize}
-          onEnterKey={state.data.refreshTitle}
+          onEnterKey={attempt.data.refreshTitle}
         />
       )}
     </Document>
   );
 }
+
+function DocumentReconnect(props: {
+  visible: boolean;
+  doc: types.DocumentTerminal;
+  attempt: Attempt<unknown>;
+  reconnect: () => void;
+}) {
+  const { message, buttonText } = getReconnectCopy(props.doc);
+
+  return (
+    <Document visible={props.visible} flexDirection="column" pl={2}>
+      <Flex
+        gap={4}
+        flexDirection="column"
+        mx="auto"
+        alignItems="center"
+        mt={100}
+      >
+        <Text typography="h5" color="text.primary">
+          {message}
+        </Text>
+        <Flex flexDirection="column" alignItems="center" mx="auto">
+          <Danger mb={3}>{props.attempt.statusText}</Danger>
+          <ButtonPrimary width="100px" onClick={props.reconnect}>
+            {buttonText}
+          </ButtonPrimary>
+        </Flex>
+      </Flex>
+    </Document>
+  );
+}
+
+function getReconnectCopy(doc: types.DocumentTerminal) {
+  switch (doc.kind) {
+    case 'doc.terminal_tsh_node': {
+      return {
+        message: 'This SSH connection is currently offline.',
+        buttonText: 'Reconnect',
+      };
+    }
+    case 'doc.terminal_shell':
+    case 'doc.terminal_tsh_kube': {
+      return {
+        message: 'Ran into an error when starting the terminal session.',
+        buttonText: 'Retry',
+      };
+    }
+    default:
+      assertUnreachable(doc);
```

Comment:
As I said before, I keep going back and forth on throwing within UI code.

The ultimate problem is that if we add a new doc type and then the user downgrades the app while having the new doc in the state, this code will crash the app. It'd be completely okay to keep it _if_ we had zod throw out any unsupported docs from app_state.json before loading them into the actual app state (which we currently don't do).

In the end, I think I prefer type safety in code vs being careful when adding new docs vs not having type safety. As in, with the current code the type system will return an error if someone extends `DocumentTerminal` without addressing these exhaustive checks.

Reply:
Are you sure it would crash the app? Isn't it that an unsupported document is caught by `DocumentRenderer`?
I mean:
```tsx
default:
        return (
          <Document visible={visible}>
            Document kind "{doc.kind}" is not supported
          </Document>
        );
```

Reply:
> Are you sure it would crash the app? Isn't it that an unsupported document is caught by `DocumentRenderer`?

Ah, you're right!